File: src/secretsetter/SecretFetcher.cs
/* This file contains a simple wrapper for setting and fetching keyvault secrets
 * 
 * Author: Josh McIntyre
 */

using System;
using System.ComponentModel;
using System.Security;
using Azure.Identity;
using Azure.Security.KeyVault.Secrets;

public class SecretFetcher
{

    private SecretClient client;

    // Define a custom exception for failed client setup
    public class NoKeyvaultConnectionException : Exception {}
    public class NoSecretFoundException : Exception {}
    public class NoSecretSetException : Exception {}

    public SecretFetcher()
    {
        setupClient();
    }

    // Configure the Azure client
    public SecretClient setupClient()
    {
        // Get keyvault URI from the environment configuration
        var keyVaultUri = Environment.GetEnvironmentVariable("KEYVAULT_URI");
        if (keyVaultUri == null)
        {
            throw new NoKeyvaultConnectionException();
        }

        // Create the secret client with the given config
        client = new SecretClient(new Uri(keyVaultUri), new DefaultAzureCredential());

        return client;
    }

    // Get a keyvault secret by name
    // Return the found secret string value
    // Or throw NoSecretFoundException
    public string getSecret(string secretName)
    {
        try
        {
            // Fetch the secret via the Azure client,
            // and extract the secret string value
            var secretRet = client.GetSecret(secretName);
            var secret = secretRet.Value.Value;

            return secret;
        }
        catch (Azure.RequestFailedException)
        {
            throw new NoSecretFoundException();
        }
    }

    // Set a keyvault secret by name
    // Throws NoSecretSetException if the secret cannot be set
    public void setSecret(string secretName, string secret)
    {
        try
        {
            // Set the secret via the Azure client
            var secretRet = client.SetSecret(new KeyVaultSecret(secretName, secret.ToString()));
        }
        catch (Azure.RequestFailedException)
        {
            throw new NoSecretSetException();
        }
    }
}

File: src/secretsetterpy/secretfetcher.py
# This file contains a simple wrapper for fetching a keyvault secret
#
# Author: Josh McIntyre
#
import os

from azure.core.exceptions import HttpResponseError, ResourceNotFoundError
from azure.keyvault.secrets import SecretClient
from azure.identity import DefaultAzureCredential

class NoKeyvaultConnectionException(Exception): pass
class NoSecretFoundException(Exception): pass
class NoSecretSetException(Exception): pass

class SecretFetcher:

    # Setup the Azure client
    def __init__(self):
    
        self.client = None

        self._setup_client()
        
    def _setup_client(self):

        kv_uri = os.environ.get("KEYVAULT_URI", None)
        if not kv_uri:
            raise NoKeyvaultConnectionException

        self.client = SecretClient(vault_url=kv_uri, credential=DefaultAzureCredential())
    
    # Fetch the secret value
    # Return the secret value string, or
    # raise NoSecretFoundException
    def get_secret(self, secret_name):
    
        try:
            secret_ret = self.client.get_secret(secret_name)
            secret = secret_ret.value
            
            return secret
        except ResourceNotFoundError:
            raise NoSecretFoundException
            
    # Set the secret by name
    # raise NoSecretSetException
    def set_secret(self, secret_name, secret):
    
        try:
            secret_ret = self.client.set_secret(secret_name, secret)
        except HttpResponseError:
            raise NoSecretSetException